geomesa-kafka/geomesa-kafka-datastore/src/main/scala/org/locationtech/geomesa/kafka/data/KafkaFeatureWriter.scala

```diff
@@ -130,4 +135,65 @@ object KafkaFeatureWriter {
       super.write()
     }
   }
+
+  trait KafkaFeatureProducer extends Flushable {
+    def send(record: ProducerRecord[Array[Byte], Array[Byte]]): Unit
+  }
+
+  case class AutoCommitProducer(producer: Producer[Array[Byte], Array[Byte]]) extends KafkaFeatureProducer {
+    override def send(record: ProducerRecord[Array[Byte], Array[Byte]]): Unit = producer.send(record)
+    override def flush(): Unit = producer.flush()
+  }
+
+  case class KafkaTransactionState(producer: Producer[Array[Byte], Array[Byte]])
+      extends Transaction.State with KafkaFeatureProducer {
+
+    private val tx = new AtomicReference[Transaction](null)
+    private var inTransaction = false
+
+    override def send(record: ProducerRecord[Array[Byte], Array[Byte]]): Unit = {
```

Comment:
Are these states handled by single threaded access?

Reply:
in the general use case, yes. it's definitely possible to use them in a multi-threaded manner, i.e. by creating a transaction and then passing it to different threads. but that seems like something that probably wouldn't be supported in geotools anyway? i thought about making this thread safe but decided not to.